Ryan tun init

prairir opened the pull request:
### Did you read the rules to contributing in `CONTRIBUTING.md`
sure

### What are you trying to accomplish?
Creating the tun interface and setting it up properly

### How are you accomplishing it?
setting up the actual tun part of the tun interface was done before. This is just using networking magic to setup the actual interface with the configuration that we setup

### Is there anything reviewers should know?
its all commented nicely

### If needed, did you change documentation(`README.md`, `ARCHITECTURE.md`, etc) to reflect your changes?
I uhhhh forgot, il do it next PR
- [x] Is it safe to rollback this change if anything goes wrong?

prairir committed:
CLI baby setup

renamed the `cmd` package to `shanty` because nautical theme and
cool. Set this stuff up with `spf13/cobra` to handle all the CLI magic
that goes on.

prairir committed:
Basic flags, Changed group name in license

Added basic flags. Add terminology section to README. Named the group
buoy boyz in the license

prairir committed:
Add: system init config

System init config needs to read in the config file, flags, and env vars
but it would be different than the system config. The actual system
config needs to be lean, tidy, and well designed because many aspects of
it need to be shared and theres no real way around that.

This init config maps all the config fields to a struct and has a nice
unmarshal function that sets global instance. Because this is basically
just used for system basics like log level, it doesnt need have a mutex
because it wont be written to after `config.UnmarshalInitConfig`

prairir committed:
Add: root cmd with init config logic

Add base cmd with config handling logic and nice logging. The init
config is for config file, env var, and flags. Because the actual shared
config between go routines is different and needs to come after tun
interface init, this has been broken up into init config and shared
config.

The logic for init config is initializing the logger. For a logger I
chose zerolog because its fast af and has a nice builder API which I
like. unfortunately the documentation is uhhh *weird*, it leaves out
some things.

For zerolog, I have set it up so the global logger is thread safe.
Because we will either have really pretty console logger which is thread
safe and not performant or writing to /var/log/buoy.log which is in
os.APPEND mode so the writes are thread safe, anway because of those our
global logger is thread safe.

This also holds the logic for setting log levels, where to write to,
unmarshalling init config, and reading config from /etc/buoy/buoy.yaml.

Also updating go version and adding dependancies like logging and cobra
and stuff.

prairir committed:
Add: example config

prairir committed:
Add: Shared config struct and functions

There needs to be a shared config of the actual running system. It has
to be thread safe, performant(enough), with nice errors. This adds most
of those.

It doesnt hold all needed fields like fleet peer list. It uses a RW
mutex to ensure that reading is async but writing is sync. This is
really a **WIP** but a general base is useful and needed

prairir committed:
Merge pull request #1 from prairir/ryan-CLI

CLI, config, and Logger

prairir committed:
Add: initialization and creation of tun interface

A few things. So it turns out the actual address in a CIDR range is
important, guess I forgot that. Either way, to actually keep it, I
changed the config to unmarshal differently. I was running with the hook
to decode the CIDR and pull out the range but as I was saying, we need
the address too. I couldnt figure out a solution using mapstructure
cause Im lazy and its 2 AM.

The solution I picked works just fine. It unmarshals as a string and
then parses out afterwards.

The tun interface stuff is pretty cool ngl. Either way, instead of doing
it manually, like we were doing before, we dont do it manually. I wrote
in comments what each part does but basically it creates the interface
with a name, makes its range and address, brings it up, connects the tun
interface to it, and then returns the interface.

eventually we should make a route that destroys the interface.

prairir commented:
dont review it yet, it has some weird files n stuff i didnt even realize

prairir commented:
oh and im going to put in the take down code for the interface

prairir committed:
Remove: mistake file

That file was a mistake, no idea how it got there

bachittle commented:
let me know when it's ready for review

prairir committed:
Fix: ioctl invalid config tun,Rename: Init to New

Renamed Init to New. Bailey pointed out that naming it `Init` could be
confusing, `New` is much more unambiguous.

Fix the ioctl invalid config error that comes from `water.New`. This
error was caused by a few things so lets go over em.

1. This issue was caused by me creating a network interface with tenus
first. I thought that I had to create the interface and then configure
it and then create a TUN device over it. This was wrong. Changing the
order so water creates first fixes that.

2. Creating an interface with tenus would also cause an issue. Switching
to tenus setting a known interface also fixes that.

prairir committed:
Change: better less used ip route

192.168.2.1/24 is very commonly used. The route that I changed it to is
much less commonly used.

prairir commented:
It can now be reviewed, please do and checkout my git messages for more info

bachittle commented:
I made an issue to address branch structure (https://github.com/prairir/Buoy/issues/3), this branch could be merged directly into master as this code is crucial for other branches regardless (currently master branch doesn't do anything, this branch at least does things).

bachittle commented:
I get this error when running `sudo go run main.go` on this branch. 
![image](https://user-images.githubusercontent.com/39804642/149682505-ddebb454-cd51-4404-806e-6773ce6dd923.png)

prairir commented:
> I get this error when running `sudo go run main.go` on this branch. ![image](https://user-images.githubusercontent.com/39804642/149682505-ddebb454-cd51-4404-806e-6773ce6dd923.png)

You need to either have a config at the base config location `/etc/buoy/buoy.yml` or you can specify with the `--config` flag

prairir commented:
> I made an issue to address branch structure (#3), this branch could be merged directly into master as this code is crucial for other branches regardless (currently master branch doesn't do anything, this branch at least does things).

I agree with this, I am going to change the base and then force merge when its time to merge

bachittle commented:
> > I get this error when running `sudo go run main.go` on this branch. ![image](https://user-images.githubusercontent.com/39804642/149682505-ddebb454-cd51-4404-806e-6773ce6dd923.png)
> 
> You need to either have a config at the base config location `/etc/buoy/buoy.yml` or you can specify with the `--config` flag

alrighty, we should have that in the README along with removing the stuff we don't need anymore, like the ip stuff since that's done in code now.

bachittle commented:
there should also be a sample config file so we know how to structure it, rather than having to look through the code to see how to structure the file

prairir commented:
there is a sample config but i will add it to the readme on how to use it

prairir commented:
I can add another commit with examples n stuff

prairir committed:
Update: readme, Fix: log-cli flag

README needed to be updated with the running and configuration stuff.
log-cli flag wasnt working properly, I fixed that

bachittle commented on README.md:
````diff
@@ -6,21 +6,56 @@
+run with printing
+
+```sh
+sudo ./buoy --log-cli
````
could potentially tell the user to run `go build` first to get the executable, then run the executable using this command.

prairir replied:
not an awful idea

bachittle commented on cmd/root.go:
```diff
@@ -0,0 +1,143 @@
+	cobra.CheckErr(rootCmd.Execute())
+}
+
+func init() {
```
I remember you telling me that this is potentally an anti-pattern, could comment about it just in case a better solution is thought of.

prairir replied:
It is an anti pattern but this code is generated for us by `cobra init` and so im not worried about it

bachittle reviewed (approved):
I made a few suggestions regarding documentation and comments, but nothing major. Can merge in its current state.

bachittle commented:
also 1 more suggestion: gitignore file for executable Buoy and any yaml files that aren't the example file
![image](https://user-images.githubusercontent.com/39804642/149827412-3bc660d2-a9c0-4a9a-928e-e1088fdecb25.png)

prairir committed:
Add: earth file with tests, deps, and building

prairir committed:
Add: tests, Update: readme with testing stuff

prairir commented:
> also 1 more suggestion: gitignore file for executable Buoy and any yaml files that aren't the example file ![image](https://user-images.githubusercontent.com/39804642/149827412-3bc660d2-a9c0-4a9a-928e-e1088fdecb25.png)

I will work on that right now

prairir committed:
Add: gitignore, Add: build step to readme

connersax reviewed (approved):
Don't see anything big missing. Bailey and you discussed a few things and it looks good so far.
